File: frontend/src/config/urls.js
/**
 * Centralized URL configuration for DocuSearch frontend.
 *
 * This module provides a single source of truth for all service URLs and endpoints
 * across the DocuSearch React application. It supports Vite environment variable
 * configuration and provides convenient properties and methods for building
 * service-specific URLs.
 *
 * Environment Variables (Vite):
 *   VITE_WORKER_URL: Processing worker base URL (default: http://localhost:8002)
 *   VITE_CHROMADB_URL: ChromaDB server URL (default: http://localhost:8001)
 *   VITE_RESEARCH_API_URL: Research API base URL (default: http://localhost:8004)
 *   VITE_COPYPARTY_URL: Copyparty file server URL (default: http://localhost:8000)
 *   VITE_FRONTEND_URL: Frontend application URL (default: http://localhost:42007)
 *   VITE_FRONTEND_PORT: Frontend dev server port (default: 42007)
 *   VITE_UI_PORT: UI viewer dev server port (default: 42008)
 *
 * Usage:
 *   import { serviceURLs } from '@/config/urls';
 *   const searchUrl = serviceURLs.worker.search;
 *   const imageUrl = serviceURLs.document.imageUrl('abc123', 'page-5.png');
 *
 * Migration Notes:
 *   - Replaces hardcoded URLs scattered across components
 *   - Centralizes Vite env var usage
 *   - Provides consistent URL building logic
 *   - Mirrors backend src/config/urls.py structure
 *
 * @module config/urls
 */

/**
 * Service base URLs from environment variables
 */
const BASE_URLS = {
  worker: import.meta.env.VITE_WORKER_URL || 'http://localhost:8002',
  chromadb: import.meta.env.VITE_CHROMADB_URL || 'http://localhost:8001',
  researchApi: import.meta.env.VITE_RESEARCH_API_URL || 'http://localhost:8004',
  copyparty: import.meta.env.VITE_COPYPARTY_URL || 'http://localhost:8000',
  frontend: import.meta.env.VITE_FRONTEND_URL || 'http://localhost:42007',
};

/**
 * Join URL path segments safely, handling trailing/leading slashes
 * @param {string} base - Base URL
 * @param {string} path - Path to append
 * @returns {string} Joined URL
 */
const urlJoin = (base, path) => {
  const baseClean = base.endsWith('/') ? base.slice(0, -1) : base;
  const pathClean = path.startsWith('/') ? path : `/${path}`;
  return `${baseClean}${pathClean}`;
};

/**
 * Worker service endpoints
 */
const workerEndpoints = {
  base: BASE_URLS.worker,
  search: urlJoin(BASE_URLS.worker, '/search'),
  status: urlJoin(BASE_URLS.worker, '/status'),
  health: urlJoin(BASE_URLS.worker, '/health'),
  documents: urlJoin(BASE_URLS.worker, '/documents'),
  documentStructure: urlJoin(BASE_URLS.worker, '/document-structure'),
};

/**
 * ChromaDB endpoints
 */
const chromadbEndpoints = {
  base: BASE_URLS.chromadb,
  heartbeat: urlJoin(BASE_URLS.chromadb, '/api/v1/heartbeat'),
  collections: urlJoin(BASE_URLS.chromadb, '/api/v1/collections'),
};

/**
 * Research API endpoints
 */
const researchApiEndpoints = {
  base: BASE_URLS.researchApi,
  ask: urlJoin(BASE_URLS.researchApi, '/api/research/ask'),
  health: urlJoin(BASE_URLS.researchApi, '/api/health'),
  status: urlJoin(BASE_URLS.researchApi, '/api/status'),
};

/**
 * Document asset URL builders
 */
const documentAssets = {
  /**
   * Build URL for a document image (page/chunk image)
   * @param {string} docId - Document ID (SHA-256 hash)
   * @param {string} imageFilename - Image filename (e.g., "page-5.png")
   * @param {boolean} [absolute=true] - Return absolute URL vs relative path
   * @returns {string} URL to document image
   *
   * @example
   * documentAssets.imageUrl('abc123', 'page-5.png')
   * // => 'http://localhost:8000/uploads/abc123/images/page-5.png'
   */
  imageUrl: (docId, imageFilename, absolute = true) => {
    const path = `/uploads/${docId}/images/${imageFilename}`;
    return absolute ? urlJoin(BASE_URLS.copyparty, path) : path;
  },

  /**
   * Build URL for a document thumbnail (cover/first page)
   * @param {string} docId - Document ID (SHA-256 hash)
   * @param {boolean} [absolute=true] - Return absolute URL vs relative path
   * @returns {string} URL to document thumbnail
   *
   * @example
   * documentAssets.thumbnailUrl('abc123')
   * // => 'http://localhost:8000/uploads/abc123/thumbnail.png'
   */
  thumbnailUrl: (docId, absolute = true) => {
    const path = `/uploads/${docId}/thumbnail.png`;
    return absolute ? urlJoin(BASE_URLS.copyparty, path) : path;
  },

  /**
   * Build URL for audio cover art
   * @param {string} docId - Document ID (SHA-256 hash)
   * @param {boolean} [absolute=true] - Return absolute URL vs relative path
   * @returns {string} URL to cover art
   *
   * @example
   * documentAssets.coverArtUrl('abc123')
   * // => 'http://localhost:8000/uploads/abc123/cover-art.png'
   */
  coverArtUrl: (docId, absolute = true) => {
    const path = `/uploads/${docId}/cover-art.png`;
    return absolute ? urlJoin(BASE_URLS.copyparty, path) : path;
  },

  /**
   * Build URL for original uploaded file
   * @param {string} docId - Document ID (SHA-256 hash)
   * @param {string} filename - Original filename
   * @param {boolean} [absolute=true] - Return absolute URL vs relative path
   * @returns {string} URL to original file
   *
   * @example
   * documentAssets.originalFileUrl('abc123', 'document.pdf')
   * // => 'http://localhost:8000/uploads/abc123/document.pdf'
   */
  originalFileUrl: (docId, filename, absolute = true) => {
    const path = `/uploads/${docId}/${filename}`;
    return absolute ? urlJoin(BASE_URLS.copyparty, path) : path;
  },
};

/**
 * Frontend route URL builders
 */
const frontendRoutes = {
  /**
   * Build document details URL with optional navigation parameters
   * @param {string} docId - Document ID (SHA-256 hash)
   * @param {Object} [options={}] - Navigation options
   * @param {number} [options.page] - Page number (1-indexed, for visual documents)
   * @param {string} [options.chunkId] - Chunk ID for navigation
   * @returns {string} Relative URL to document details page
   *
   * @example
   * frontendRoutes.detailsUrl('abc123', { page: 5 })
   * // => '/details/abc123?page=5'
   *
   * frontendRoutes.detailsUrl('abc123', { chunkId: 'abc123-chunk0042' })
   * // => '/details/abc123?chunk=abc123-chunk0042'
   */
  detailsUrl: (docId, options = {}) => {
    if (!docId) {
      throw new Error('docId is required');
    }

    let path = `/details/${docId}`;

    const params = new URLSearchParams();
    if (options.page && typeof options.page === 'number' && options.page > 0) {
      params.set('page', String(options.page));
    }
    if (options.chunkId && typeof options.chunkId === 'string' && options.chunkId.trim()) {
      params.set('chunk', options.chunkId.trim());
    }

    const queryString = params.toString();
    if (queryString) {
      path = `${path}?${queryString}`;
    }

    return path;
  },

  /**
   * Build search results URL with query
   * @param {string} query - Search query
   * @returns {string} Relative URL to search results
   *
   * @example
   * frontendRoutes.searchUrl('machine learning')
   * // => '/search?q=machine+learning'
   */
  searchUrl: (query) => {
    if (!query) {
      return '/search';
    }
    return `/search?q=${encodeURIComponent(query)}`;
  },
};

/**
 * Centralized service URLs configuration
 *
 * Provides organized access to all service endpoints and URL builders.
 *
 * @example
 * import { serviceURLs } from '@/config/urls';
 *
 * // Worker endpoints
 * fetch(serviceURLs.worker.search, { ... });
 *
 * // Document assets
 * const imgSrc = serviceURLs.document.imageUrl(docId, 'page-5.png');
 *
 * // Frontend routes
 * navigate(serviceURLs.frontend.detailsUrl(docId, { page: 5 }));
 */
export const serviceURLs = {
  worker: workerEndpoints,
  chromadb: chromadbEndpoints,
  researchApi: researchApiEndpoints,
  document: documentAssets,
  frontend: frontendRoutes,
  base: BASE_URLS,
};

/**
 * Export individual endpoint groups for convenience
 */
export { workerEndpoints, chromadbEndpoints, researchApiEndpoints, documentAssets, frontendRoutes };

/**
 * Export base URLs for direct access
 */
export { BASE_URLS };

/**
 * Default export for backward compatibility
 */
export default serviceURLs;

File: src/config/urls.py
"""
Centralized URL configuration for DocuSearch backend services.

This module provides a single source of truth for all service URLs and endpoints
across the DocuSearch application. It supports environment variable configuration
and provides convenient properties and methods for building service-specific URLs.

Environment Variables:
    WORKER_URL: Processing worker base URL (default: http://localhost:8002)
    RESEARCH_API_URL: Research API base URL (default: http://localhost:8004)
    FRONTEND_URL: Frontend application URL (default: http://localhost:3333)

Usage:
    >>> from src.config.urls import get_service_urls
    >>> urls = get_service_urls()
    >>> search_url = urls.worker_search

Thread Safety:
    This module uses a singleton pattern with thread-local storage to ensure
    thread-safe access to URL configuration.
"""

import logging
import os
from typing import Optional
from urllib.parse import urljoin

logger = logging.getLogger(__name__)


class ServiceURLs:
    """
    Centralized configuration for all service URLs and endpoints.

    This class provides properties for common endpoints and methods for building
    dynamic URLs with document-specific paths.

    Attributes:
        worker: Processing worker base URL
        research_api: Research API base URL
        frontend: Frontend application base URL
    """

    def __init__(self) -> None:
        """Initialize service URLs from environment variables."""
        # Service base URLs
        self.worker = os.getenv("WORKER_URL", "http://localhost:8002")
        self.research_api = os.getenv("RESEARCH_API_URL", "http://localhost:8004")
        self.frontend = os.getenv("FRONTEND_URL", "http://localhost:3333")

        # Log configuration on initialization
        logger.debug(
            f"Initialized ServiceURLs: worker={self.worker}, "
            f"research_api={self.research_api}, frontend={self.frontend}"
        )

    # =========================================================================
    # Worker Service Endpoints
    # =========================================================================

    @property
    def worker_search(self) -> str:
        """Worker search endpoint URL."""
        return urljoin(self.worker, "/search")

    @property
    def worker_status(self) -> str:
        """Worker status endpoint URL."""
        return urljoin(self.worker, "/status")

    @property
    def worker_health(self) -> str:
        """Worker health check endpoint URL."""
        return urljoin(self.worker, "/health")

    @property
    def worker_documents(self) -> str:
        """Worker documents endpoint URL."""
        return urljoin(self.worker, "/documents")

    @property
    def worker_document_structure(self) -> str:
        """Worker document structure endpoint URL."""
        return urljoin(self.worker, "/document-structure")

    # =========================================================================
    # Research API Endpoints
    # =========================================================================

    @property
    def research_ask(self) -> str:
        """Research API ask endpoint URL."""
        return urljoin(self.research_api, "/api/research/ask")

    @property
    def research_health(self) -> str:
        """Research API health check endpoint URL."""
        return urljoin(self.research_api, "/api/health")

    @property
    def research_status(self) -> str:
        """Research API status endpoint URL."""
        return urljoin(self.research_api, "/api/status")

    # =========================================================================
    # Frontend URLs
    # =========================================================================

    def frontend_details_url(
        self, doc_id: str, page: Optional[int] = None, chunk_id: Optional[str] = None
    ) -> str:
        """
        Build frontend document details URL with optional navigation parameters.

        Note: For more advanced URL building with relative/absolute control,
        use src.utils.url_builder.build_details_url() directly.

        Args:
            doc_id: Document ID (SHA-256 hash)
            page: Optional page number (1-indexed, for visual documents)
            chunk_id: Optional chunk ID for navigation

        Returns:
            Absolute URL to document details page in frontend

        Example:
            >>> urls = ServiceURLs()
            >>> urls.frontend_details_url("abc123", page=5)
            'http://localhost:3333/details/abc123?page=5'
        """
        from src.utils.url_builder import build_details_url

        return build_details_url(doc_id=doc_id, page=page, chunk_id=chunk_id, absolute=True)


# =============================================================================
# Singleton Pattern with Thread-Local Storage
# =============================================================================

import threading

_thread_local = threading.local()


def get_service_urls() -> ServiceURLs:
    """
    Get the singleton ServiceURLs instance for the current thread.

    This function implements a thread-safe singleton pattern, ensuring that
    each thread gets its own ServiceURLs instance while avoiding race conditions
    in multi-threaded environments.

    Returns:
        ServiceURLs instance for the current thread

    Example:
        >>> from src.config.urls import get_service_urls
        >>> urls = get_service_urls()
        >>> search_url = urls.worker_search
    """
    if not hasattr(_thread_local, "service_urls"):
        _thread_local.service_urls = ServiceURLs()
        logger.debug("Created new ServiceURLs instance for thread")
    # Type assertion for mypy - we just created it above if it didn't exist
    result: ServiceURLs = _thread_local.service_urls
    return result


def reset_service_urls() -> None:
    """
    Reset the ServiceURLs singleton for the current thread.

    Useful for testing or when environment variables change at runtime.

    Example:
        >>> import os
        >>> os.environ["WORKER_URL"] = "http://newworker:8002"
        >>> reset_service_urls()
        >>> urls = get_service_urls()  # Will use new WORKER_URL
    """
    if hasattr(_thread_local, "service_urls"):
        delattr(_thread_local, "service_urls")
        logger.debug("Reset ServiceURLs instance for thread")
